src/main/java/com/book/service/ProductsService.java

```diff
@@ -88,4 +100,45 @@ public Optional<ProductsEntity> findProductById(int id) {
     public Optional<Page<ProductsEntity>> findAllByProductNameContainingIgnoreCase(int page, int size, String productName) {
         return Optional.ofNullable(productsRepository.findAllByProductNameContainingIgnoreCase(PageRequest.of(page, size), productName));
     }
+
+    @Override
+    @Transactional
+    public Optional<List<ProductsEntity>> getRecommendedProducts(int userId) throws IOException {
+        Optional<List<ProductReviewsEntity>> productReviewsEntityList = productReviewsService.findAll();
+        List<List<Integer>> allRatingList = new ArrayList<>();
+
+        if (productReviewsEntityList.isEmpty()) {
+            return Optional.empty();
+        }
+
+        for(ProductReviewsEntity productReviewsEntity: productReviewsEntityList.get()) {
+            List<Integer> temp = new ArrayList<>();
+            temp.add(productReviewsEntity.getUsersEntity().getId());
+            temp.add(productReviewsEntity.getProductId());
+            temp.add(productReviewsEntity.getRating());
+
+            allRatingList.add(temp);
+        }
+
+        for (int i = 0; i < allRatingList.size(); i++) {
```

Comment:
i'm not sure what this for loop is doing?

Reply:
this for loop is used to calculate the average rating of an user gave to a product. This is because i get all the ratings from the database, and an user is able to give more than 1 rating to a product. 
Also, the python script requires an array in this format [userid, productid, rating].

Reply:
this code block is hard to understand because u used a 2-dimensions array `List<List<Integer>> allRatingList`
A more clean-code way is to define a `RatingObject[userid, productid, rating]`, and then declare List<RatingObject> allRatingList, then using Java Streaming API to calculate average: https://stackoverflow.com/questions/50619213/group-and-calculate-average-of-object-array-using-stream-in-java